homeassistant/components/mqtt/fan.py

```diff
@@ -69,40 +70,56 @@
 
 OSCILLATION = "oscillation"
 
-PLATFORM_SCHEMA = (
-    mqtt.MQTT_RW_PLATFORM_SCHEMA.extend(
-        {
-            vol.Optional(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
-            vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-            vol.Optional(CONF_OPTIMISTIC, default=DEFAULT_OPTIMISTIC): cv.boolean,
-            vol.Optional(CONF_OSCILLATION_COMMAND_TOPIC): mqtt.valid_publish_topic,
-            vol.Optional(CONF_OSCILLATION_STATE_TOPIC): mqtt.valid_subscribe_topic,
-            vol.Optional(CONF_OSCILLATION_VALUE_TEMPLATE): cv.template,
-            vol.Optional(CONF_PAYLOAD_HIGH_SPEED, default=SPEED_HIGH): cv.string,
-            vol.Optional(CONF_PAYLOAD_LOW_SPEED, default=SPEED_LOW): cv.string,
-            vol.Optional(CONF_PAYLOAD_MEDIUM_SPEED, default=SPEED_MEDIUM): cv.string,
-            vol.Optional(CONF_PAYLOAD_OFF_SPEED, default=SPEED_OFF): cv.string,
-            vol.Optional(CONF_PAYLOAD_OFF, default=DEFAULT_PAYLOAD_OFF): cv.string,
-            vol.Optional(CONF_PAYLOAD_ON, default=DEFAULT_PAYLOAD_ON): cv.string,
-            vol.Optional(
-                CONF_PAYLOAD_OSCILLATION_OFF, default=OSCILLATE_OFF_PAYLOAD
-            ): cv.string,
-            vol.Optional(
-                CONF_PAYLOAD_OSCILLATION_ON, default=OSCILLATE_ON_PAYLOAD
-            ): cv.string,
-            vol.Optional(CONF_SPEED_COMMAND_TOPIC): mqtt.valid_publish_topic,
-            vol.Optional(
-                CONF_SPEED_LIST,
-                default=[SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH],
-            ): cv.ensure_list,
-            vol.Optional(CONF_SPEED_STATE_TOPIC): mqtt.valid_subscribe_topic,
-            vol.Optional(CONF_SPEED_VALUE_TEMPLATE): cv.template,
-            vol.Optional(CONF_STATE_VALUE_TEMPLATE): cv.template,
-            vol.Optional(CONF_UNIQUE_ID): cv.string,
-        }
+
+def _speed_list_validator(config):
+    """Validate the speed list."""
+    config = copy.deepcopy(config)
+
+    if CONF_SPEED_COMMAND_TOPIC not in config or CONF_SPEED_STATE_TOPIC not in config:
+        config[CONF_SPEED_LIST] = None
```

Comment:
Shouldn't we instead require speed command or speed state topic if speed list is set? Ie, make these inclusive.

As a user I would be confused if I configure the speed list but it is reset to `None` without notice.

Reply:
That sounds like a great idea. How could I achieve the requirement for speed command/state topic if speed list is set?

Reply:
Add to the check here to check for speed list presence. And if the check is true, raise `vol.Invalid`.

Reply:
Or maybe invert the check and return the config if true otherwise raise. It depends on what is easier to write. If speed list is present any of command or state topic must be set, otherwise raise.

Reply:
That sounds very compelling but I am not sure how this fixes this issue. If speed state topic and speed command topic is not set, your described condtion will be true nevertheless since the default value for the speed list is automatically set. Therefore, I did not add a check to test the presence of the speed list since it is always set.

Reply:
Yeah, it's weird that we have a default for the speed list.

I looked more at the issue description and I think the problem is that we don't set supported features properly and/or the frontend doesn't take supported features into account when deciding to show the drop down.

We should support set speed if we have a command topic. That is required to set speed. State topic is optional as it just decides if we get feedback for state update.

Showing the speed list drop down or not should be based on supported features in the frontend as far as I know.

Reply:
Yeah, I agree it looks like the frontend does not take the supported_features into account when showing the speed list drop down. 

> We should support set speed if we have a speed command topic. That is required to set speed. Speed state topic is optional as it just decides if we get feedback for state update.

In the code it is slightly different, should this be changed?
https://github.com/home-assistant/home-assistant/blob/e793ed9ab0b8857ba0aa88a7e7bfb8accd7761b2/homeassistant/components/mqtt/fan.py#L237-L239

Reply:
Yes, that's wrong.

Reply:
Changed with https://github.com/home-assistant/home-assistant/pull/28680/commits/e209cb408c5c444d42988283cef528f280780737

> Showing the speed list drop down or not should be based on supported features in the frontend as far as I know.

I have glanced over the frontend implementation of this more-info-fan card and it looks like it does not use the supported_features at all and instead loops through the speed list. Do you know of an example where the supported_features are used in the frontend? Or do you know how to implement it correctly?

Reply:
Sorry, I'm not very deep into the frontend. It's probably best to check with the frontend team, eg in discord.

Reply:
Okay thanks for the tip 👍
Should I revert https://github.com/home-assistant/home-assistant/pull/28680/commits/f000396fa642c64bde40513ea70d9915dbd71ead so this can be merged?

Reply:
Yes, please! :+1: